File: Data_structure/841/hash.cpp
#include <iostream>
using namespace std;

const int N = 1e5 + 10, P = 131;
int n, m, l1, r1, l2, r2;
char s[N];
unsigned long long h[N], p[N];

unsigned long long find_hash(int l, int r);

int main() {
    scanf("%d%d%s", &n, &m, s + 1);
    p[0] = 1;
    for (int i = 1; i <= n; i++) {
        p[i] = p[i - 1] * P;
        h[i] = h[i - 1] * P + s[i] - '0' + 1;
    }
    while (m--) {
        scanf("%d%d%d%d", &l1, &r1, &l2, &r2);
        if (find_hash(l1, r1) == find_hash(l2, r2)) printf("Yes\n");
        else printf("No\n");
    }
    return 0;
}

unsigned long long find_hash(int l, int r) {
    return h[r] - h[l - 1] * p[r - l + 1];
}

File: Data_structure/841/hash.py
N, P = 2 ** 64, 131
def find_hash(l: int, r: int, h: list, p: list) -> int:
    return (h[r] - h[l - 1] * p[r - l + 1]) % N


if __name__ == "__main__":
    n, m = map(int, input().split())
    h, p = [0] * (n + 1), [1] * (n + 1)
    s = input()
    for i in range(len(s)):
        p[i + 1] = p[i] * P % N
        h[i + 1] = (h[i] * P + ord(s[i]) - ord('0') + 1) % N
    for _ in range(m):
        l1, r1, l2, r2 = map(int, input().split())
        if (find_hash(l1, r1, h, p) == find_hash(l2, r2, h, p)):
            print("Yes")
        else:
            print("No")
